packages/network-controller/src/NetworkController.ts

```diff
@@ -896,6 +986,234 @@ export class NetworkController extends BaseControllerV2<
+   * Converts the provider config object in state to a network client
+   * configuration object.
+   *
+   * @returns The network client config.
+   * @throws If the provider config is of type "rpc" and lacks either a
+   * `chainId` or an `rpcUrl`.
+   */
+  #buildIdentifiedNetworkClientConfigurationFromProviderConfig(): [
+    string,
+    NetworkClientConfiguration,
+  ] {
+    const { providerConfig } = this.state;
+
+    if (providerConfig.type === NetworkType.rpc) {
+      if (providerConfig.chainId === undefined) {
+        throw new Error('chainId must be provided for custom RPC endpoints');
+      }
+      if (providerConfig.rpcUrl === undefined) {
+        throw new Error('rpcUrl must be provided for custom RPC endpoints');
+      }
+      const id =
+        providerConfig.id ??
+        [providerConfig.chainId, providerConfig.rpcUrl].join('||');
+      const networkClientConfiguration: CustomNetworkClientConfiguration = {
+        chainId: providerConfig.chainId,
+        rpcUrl: providerConfig.rpcUrl,
+        type: NetworkClientType.Custom,
+      };
+      return [id, networkClientConfiguration];
+    }
+
+    if (isInfuraProviderType(providerConfig.type)) {
+      const id = providerConfig.type;
+      const networkClientConfiguration: InfuraNetworkClientConfiguration = {
+        network: providerConfig.type,
+        infuraProjectId: this.#infuraProjectId,
+        type: NetworkClientType.Infura,
+      };
+      return [id, networkClientConfiguration];
+    }
+
+    throw new Error(`Unrecognized network type: '${providerConfig.type}'`);
+  }
+
+  /**
+   * Uses the information in the provider config object to look up a known and
+   * preinitialized network client. Once a network client is found, updates the
+   * provider and block tracker proxy to point to those from the network client,
+   * then finally creates an EthQuery that points to the provider proxy.
+   *
+   * @throws If no network client could be found matching the current provider
+   * config.
+   */
+  #applyNetworkSelection() {
+    if (!this.#autoManagedNetworkClientRegistry) {
+      throw new Error(
+        'initializeProvider must be called first in order to switch the network',
+      );
+    }
+
+    const { providerConfig } = this.state;
+
+    if (
+      providerConfig.type === NetworkType.rpc &&
+      providerConfig.id === undefined
+    ) {
+      if (providerConfig.chainId === undefined) {
+        throw new Error('chainId must be provided for custom RPC endpoints');
+      }
+      if (providerConfig.rpcUrl === undefined) {
+        throw new Error('rpcUrl must be provided for custom RPC endpoints');
+      }
+    }
+
+    const networkClientType =
+      providerConfig.type === NetworkType.rpc
+        ? NetworkClientType.Custom
+        : NetworkClientType.Infura;
+    const networkClientId =
+      providerConfig.type === NetworkType.rpc
+        ? providerConfig.id ??
+          [providerConfig.chainId, providerConfig.rpcUrl].join('||')
```

Comment:
Nit: I see this "generate ID from unrecognized provider config" step is listed in two separate places. Could we move it to a common function?

Reply:
I extracted this and made some other refactors here: https://github.com/MetaMask/core/pull/1439/commits/c1dfb6b7d70067b68e3a690b335759fc0d3b7eb3